pkg/process/procutil/dirent_linux.go

```diff
@@ -0,0 +1,100 @@
+// +build linux
+
+package procutil
+
+import (
+	"bytes"
+	"syscall"
+	"unsafe"
+)
+
+var (
+	dotBytes       = []byte(".")
+	doubleDotBytes = []byte("..")
+)
+
+// countDirent parses directory entries in buf and return the number of entries
+func countDirent(buf []byte) (consumed int, count int) {
+	origlen := len(buf)
+	count = 0
+	for len(buf) > 0 {
+		reclen, ok := direntReclen(buf)
+		if !ok || reclen > uint64(len(buf)) {
+			return origlen, count
+		}
+
+		rec := buf[:reclen]
+		buf = buf[reclen:]
+
+		ino, ok := direntIno(rec)
+		if !ok {
+			break
+		}
+		if ino == 0 { // File absent in directory.
+			continue
+		}
+		const namoff = uint64(unsafe.Offsetof(syscall.Dirent{}.Name))
+		namlen, ok := direntNamlen(rec)
+		if !ok || namoff+namlen > uint64(len(rec)) {
+			break
+		}
+		name := rec[namoff : namoff+namlen]
+		for i, c := range name {
+			if c == 0 {
+				name = name[:i]
+				break
+			}
+		}
+
+		if bytes.Equal(name, dotBytes) || bytes.Equal(name, doubleDotBytes) {
+			// Check for useless names before allocating a string.
+			continue
+		}
+		count++
+	}
+
+	return origlen - len(buf), count
+}
+
+func direntReclen(buf []byte) (uint64, bool) {
+	return readInt(buf, unsafe.Offsetof(syscall.Dirent{}.Reclen), unsafe.Sizeof(syscall.Dirent{}.Reclen))
+}
+
+func direntIno(buf []byte) (uint64, bool) {
+	return readInt(buf, unsafe.Offsetof(syscall.Dirent{}.Ino), unsafe.Sizeof(syscall.Dirent{}.Ino))
+}
+
+func direntNamlen(buf []byte) (uint64, bool) {
+	reclen, ok := direntReclen(buf)
+	if !ok {
+		return 0, false
+	}
+	return reclen - uint64(unsafe.Offsetof(syscall.Dirent{}.Name)), true
+}
+
+// readInt returns the size-bytes unsigned integer in native byte order at offset off.
+func readInt(b []byte, off, size uintptr) (u uint64, ok bool) {
+	if len(b) < int(off+size) {
+		return 0, false
+	}
+	return readIntLE(b[off:], size), true
```

Comment:
I added back. Initially I was thinking we could get away with just little endian because it covers most of the cases, but I added everything back to properly address this and cover all cases just as what the source code does.